src/libtilt/projection/project_real.py

```diff
@@ -75,3 +75,71 @@ def _project_volume(rotation_matrix) -> torch.Tensor:
     xl, xh = padding[2, 0], -padding[2, 1]
     images = [_project_volume(matrix)[yl:yh, xl:xh] for matrix in rotation_matrices]
     return torch.stack(images, dim=0)
+
+
+def project_real_2d(image: torch.Tensor, rotation_matrices: torch.Tensor) -> torch.Tensor:
+    """Make 1D projections of a 2D image in specific orientations.
+
+    Projections are made by
+
+    1. generating a line of coordinates sufficient to cover
+       the image in any orientation.
+
+    2. left-multiplying `rotation matrices` and coordinate grids to
+       produce rotated coordinates.
+
+    3. sampling `image` at rotated coordinates.
+
+    4. summing samples along depth dimension of a `(h, w)` image.
+
+    The rotation center of `image` is taken to be `torch.tensor(image.shape) // 2`.
+
+    Parameters
+    ----------
+    image: torch.Tensor
+        `(h, w)` image from which projections will be made.
+    rotation_matrices: torch.Tensor
+        `(batch, 2, 2)` array of rotation matrices
+
+    Returns
+    -------
+    projection_lines: torch.Tensor
+        `(batch, w)` array of 1D projection lines sampled from `image`.
+    """
+    image = torch.as_tensor(image)
+    rotation_matrices = torch.as_tensor(rotation_matrices, dtype=torch.float)
+    image_shape = torch.tensor(image.shape)
+    ps = padded_sidelength = int(3 ** 0.5 * torch.max(image_shape))
+    shape_difference = torch.abs(padded_sidelength - image_shape)
+    padding = torch.empty(size=(2, 2), dtype=torch.int16)
+    padding[:, 0] = torch.div(shape_difference, 2, rounding_mode='floor')
+    padding[:, 1] = shape_difference - padding[:, 0]
+    torch_padding = torch.flip(padding, dims=(0,))  # hw -> wh for torch.nn.functional.pad
+    torch_padding = einops.rearrange(torch_padding, 'wh pad -> (wh pad)')
+    image = F.pad(image, pad=tuple(torch_padding), mode='constant', value=0)
+    padded_image_shape = (ps, ps)
+    image_coordinates = coordinate_grid(image_shape=padded_image_shape)
+    image_coordinates -= padded_sidelength // 2  # (h, w, yx)
+    image_coordinates = torch.flip(image_coordinates, dims=(-1,))  # (h, w, yx)
+    image_coordinates = einops.rearrange(image_coordinates, 'h w yx -> h w yx 1')
+
+    def _project_image(rotation_matrix) -> torch.Tensor:
+        rotated_coordinates = rotation_matrix @ image_coordinates
+        rotated_coordinates += padded_sidelength // 2
+        rotated_coordinates = einops.rearrange(rotated_coordinates, 'h w yx 1 -> 1 h w yx')
+        rotated_coordinates = torch.flip(rotated_coordinates, dims=(-1,))  # yx -> yx
```

Comment:
```suggestion
        rotated_coordinates = torch.flip(rotated_coordinates, dims=(-1,))  # yx -> xy
```

think this is correct, can you confirm?

Reply:
you're right! and for the volume also updated